pylsp/workspace.py

```diff
@@ -152,6 +167,9 @@ def _progress_begin(
         percentage: Optional[int] = None,
     ) -> str:
         token = str(uuid.uuid4())
+
+        self._endpoint.request(self.M_INITIALIZE_PROGRESS, {'token': token}).result()
```

Comment:
This is not being awaited, is it?

The server must wait for the response and in case of an error not send progress reports.

Reply:
that's what `.result` does.

Reply:
https://docs.python.org/3.11/library/concurrent.futures.html#concurrent.futures.Future.result

Reply:
and https://github.com/python-lsp/python-lsp-jsonrpc/blob/c73fbdba2eeb99b7b145dcda76e62250552feda4/pylsp_jsonrpc/endpoint.py#L60-L87

Reply:
I see. It's hard to know that since many of the types in this file are missing/Unknown.

That said, I see a type issue that this change introduces.

In `python_lsp.py` the `self._endpoint` should be types as `Endpoint` and it is on `develop` branch:

![Screenshot 2023-01-02 at 20 54 10](https://user-images.githubusercontent.com/153197/210273704-b35a6563-a73f-4211-9dca-54b94ae8d4bf.png)

But with those changes it's `Unknown | Endpoint`:

![Screenshot 2023-01-02 at 20 56 46](https://user-images.githubusercontent.com/153197/210273736-de7540cf-3726-4472-803e-1095c0338805.png)

Since this code base is not very strict on type checking, it doesn't show up as linting issue but IMO it should still be fixed.

Reply:
Looked more into this file and nothing is really annotated with types so this issue probably doesn't matter. It should probably be a separate task to get types right in this server. So feel free to resolve this thread (I can't).